### ISSUE TITLE
CoqIDE is slow on OS X

### ISSUE THREAD

**Issue by julesjacobs** [coq/coq#11178]
#### Description of the problem

CoqIDE is much slower on OS X than on Linux. The cursor moves slowly when using the arrow keys, and there is a delay between typing and the text appearing on the screen. Stepping through proofs is slower as well.

#### Coq Version

The Coq Proof Assistant, version 8.10.1 (November 2019)
compiled on Nov 4 2019 15:04:04 with OCaml 4.09.0

**Comment by julesjacobs** [comment]
Further information about this issue:

When compiling the Coq standard library or when compiling Iris, or when processing proof scripts in vscoq or coqide, coqc and coqidetop.opt use very high amounts of memory (coqc 2-8GB and coqidetop.opt 30-60GB), which makes them very slow. The problem occurs on MacOS (10.14.6).

Hence the problem is not just in CoqIDE, although the problem is most visible there. Compiling with coqc is ~10x slower than on Linux, and CoqIDE takes ~1 second to process a line of proof script, which may well be 100x slower than on Linux. However, it appears to be a general problem with Coq, since invoking coqc from the command line is also much slower than on Linux, for instance, compiling Coq itself takes an hour, and almost all of that time it spent in coqc compiling the standard library. I have tried this with Coq 8.10.1, 8.10.2 and 8.9.0. All have the same problem.

In summary, the problem is not just CoqIDE but Coq in general, and might be a memory leak.

**Comment by ejgallego** [comment]
Hi @julesjacobs , I suggest you try to `strace` the `coqc` process and see if there is some strange FS access.

**Comment by julesjacobs** [comment]
Thanks, I ran dtruss, but I don't know how to interpret the output: https://pastebin.com/iWh1BtxG

There are some errors in there ("dtrace: error on enabled probe ID 2172 (ID 161: syscall::write:return): invalid kernel access in action #12 at DIF offset 68"). Are they related to this issue?

**Comment by ejgallego** [comment]
Indeed @julesjacobs there seems to be some problem with the native mode; can you recompile Coq with native support and try? I think the easiest way to do that with opam is to use `opam pin edit coq`, then tweak configure as needed.

Be sure the coq folder is clean of native compiler stuff.

**Comment by julesjacobs** [comment]
I added the option "-native-compiler" "yes", but during the recompilation I see coqc using 35GB, so it doesn't seem to have fixed the problem. Shall I run dtruss again after the recompile is finished?

**Comment by ejgallego** [comment]
Sorry I did a huge typo, I mean "wihtout" native support, you should use `-native-compiler no`. The `dtruss` you sent was having problems with the `.coq-native` stuff indeed.

Coq makefiles are a bit flaky so indeed your problem may be leftover files from previous compilations. I'd suggest:
- use a fresh opam switch
- be sure that in the coq folder there is not `.coq_native` stuff

**Comment by julesjacobs** [comment]
Thanks, that solved the problem!! Coqc is now 10x faster and uses 100x less memory. I tried doing coqc with -native-compiler no before but apparently that was not enough, but compiling coq without native support solved it. Thanks so much!

**Comment by ejgallego** [comment]
Thanks a lot @julesjacobs for debugging the problem; indeed that's extremely worrying as this setup is the default , we have a very similar bug in https://github.com/coq/coq/issues/11107

Setting a milestone as this kind of problem may be release-critical.

**Comment by ejgallego** [comment]
One last question @julesjacobs , what OCaml variant are you using? 4.09.0 flambda?

**Comment by Blaisorblade** [comment]
@ejgallego Not flambda (I'm next desk). But I measured slowdowns effects from `-native_compiler` even on Linux and without flambda (https://github.com/coq/coq/issues/11107#issuecomment-567040281).

**Comment by julesjacobs** [comment]
I'm using 4.09 without flambda.

**Comment by Blaisorblade** [comment]
I'll add that @julesjacobs is using macOS 10.15 and I'm using macOS 10.14 (and I'm *not* upgrading yet). *Potentially*, there are *two* bugs:
- The strace/dtruss log suggests native libraries are being... loaded? (when they shouldn't), with a blink-and-you-miss-it effect.
- On Jules' machine, maybe because of some bug on macOS 10.15, the native loading is massively slower.

**Comment by Blaisorblade** [comment]
*Potentially*, there are *two* bugs:
- The strace/dtruss log suggests native libraries are being... loaded? (when they shouldn't), with a blink-and-you-miss-it effect.
- On Jules' machine, the native loading is massively slower — and that might affect native-compiler anyway.

(Unlike I wrote before, both I and @julesjacobs are using macOS 10.14.6).

**Comment by ejgallego** [comment]
Indeed, thanks @Blaisorblade , that seems a bit worrying; last time I looked I couldn't understand how they would be loaded :/

**Comment by ppedrot** [comment]
Looking at the code, it seems that passing `-native-compiler off` to `coqc` only prevents it to use it in the kernel conversion and to output the corresponding object file. It does not turn off the remaining part of the machinery that the configure flag seems to be setting.

**Comment by ejgallego** [comment]
If that's the case indeed that's a bug as loading the libs does not have trivial cost so it seems; maybe just adding a few `Lazy` would do the job?

**Comment by ppedrot** [comment]
Nah, lazy is its own can of worms. We can protect the calls via a centralized flag that turns the native compiler off. The problem seems that we have a proliferation of several boolean options regarding the use of native compilation or not, and they are not synchronized nor used uniformly.

**Comment by ejgallego** [comment]
Indeed that's an old problem @ppedrot , on the other hand having native initialization to happen under `Lazy` could work, I guess I'd depend on the state. It doesn't sound too unreasonable for me.

Keep in mind that in some cases you may want Coq to do the initialization when the native conversion is called, so that's a bit tricky to govern with a flag.

The issue here is more "only initialize if the file effectively calls native conversion"

**Comment by ppedrot** [comment]
Well if the user asks `-native-compiler off` there is no way we should initialize anything.

**Comment by ppedrot** [comment]
Despite the scary comment in the file, I think that such a global flag would belong to the `Flags` module. This is a compilation option that has no sensible synchronization behaviour and is utterly global... WDYT?

**Comment by ejgallego** [comment]
Which flag concretely @ppedrot ?

**Comment by ejgallego** [comment]
If you mean the one in the env IMHO it is in the place it should be. I don't see how that flag is "global", on the contrary, it is very specific to a few modules and of course it can be synchronized with `Set`.

**Comment by llelf** [comment]
> suggests native libraries are being... loaded?

They do! (Sorry, I actually have figured it out a month ago…and then forget to comment).

You can see that (on macOS) by starting building something and then doing (in the separate terminal):

```sh
$ sudo dtrace -q -Wcoqc -n 'pid$target::dlopen:entry { @["files loaded"]=count(); printf("%s\n", copyinstr(arg0)) }'
     ⋮
    ⟨many lines skipped⟩
     ⋮
/usr/local/packages/coq.409+FL/lib/coq/theories/Numbers/Natural/Peano/.coq-native/NCoq_Numbers_Natural_Peano_NPeano.cmxs
/usr/local/packages/coq.409+FL/lib/coq/theories/QArith/.coq-native/NCoq_QArith_Qcanon.cmxs
/usr/local/packages/coq.409+FL/lib/coq/theories/QArith/.coq-native/NCoq_QArith_Qcanon.cmxs

  files loaded                                                    446
```

So, 446 for each `coqc` invocation. And only 20 with `-native-compiler no`.

**Comment by ejgallego** [comment]
Thanks a lot for the data @llelf , we will for sure fix this in 8.11, and IMHO it would be worth to fix in 8.10 (cc: @vbgl )

**Comment by Blaisorblade** [comment]
User requirement: Please make sure to document under which condition users pay a cost for native_compiler.

Ideally users pay a cost only when you use the `native_compute` tactic, and maybe only for libraries that are actually used? It sounds like that might be hard, so if that fails, please let us users know :-).

**Comment by ppedrot** [comment]
How come linking is that expensive on MacOS compared to Linux? Or is is just the file opening that causes trouble?

**Comment by Blaisorblade** [comment]
> How come linking is that expensive on MacOS compared to Linux? Or is is just the file opening that causes trouble?

Unclear, but there clearly is a *second* bug on *some* macOS machine, which hits @julesjacobs but not me.
As I said, both @julesjacobs and I have macOS, he has a faster machine, but linking seems orders of magnitude more expensive on his machine, for unclear reasons.
Also, coqIDE seems much slower on his machine, while normal programs (including Ocaml ones) seem fine, and early troubleshooting was inconclusive.

**Comment by ppedrot** [comment]
I am postponing this, because it's not a 8.11 regression, even though it is a serious bug. Furthermore, it doesn't seem to be directly related to Coq, but rather to the linking mechanism of OCaml.

**Comment by ppedrot** [comment]
I am wondering whether this could this be linked to the filesystem used by the underlying OS. @julesjacobs and @Blaisorblade, are you using HFS+ or APFS?

**Comment by Blaisorblade** [comment]
I'm on 10.14 APFS, and don't suffer the bug significantly.

**Comment by davidnowak** [comment]
On macOS Catalina 10.15.3 with an APFS filesystem, Coqide 8.11.0 (installed with Opam) is slow (compared to 8.9.1). I cannot quantify the slowdown, but it significative enough for me to stick to 8.9.1.

**Comment by ejgallego** [comment]
@davidnowak did you try the suggested workaround [disabling native compute] ? Does it improve things?

**Comment by ppedrot** [comment]
@davidnowak does it also happen for batch compilation? You could check that compiling some arbitrary file with the `-time` option for coqc.

**Comment by davidnowak** [comment]
> @davidnowak does it also happen for batch compilation?

I have tried that on [this file](https://github.com/davidnowak/cecoa/blob/master/src/Cecoa/Lib.v). In batch compilation, the slowdown (w.r.t. 8.9.1) is 0.4%. But within Coqide the slowdown (w.r.t. 8.9.1) is 10.8%.

**Comment by davidnowak** [comment]
> @davidnowak did you try the suggested workaround [disabling native compute] ? Does it improve things?

There is no slowdown with this workaround.

**Comment by ejgallego** [comment]
Thanks for confirming @davidnowak , so this is a duplicate it seems; would be great to fix that soon and release 8.11.1

**Comment by ppedrot** [comment]
@ejgallego how do we fix that though? Should we merely implement the lazy linking of native compute binary objects or fix the OCaml linker? The first solution looks more like a workaround to me, while the second will require cooperation with upstream OCaml developers.

**Comment by ejgallego** [comment]
I thought we had a PR fixing that; I'd say the problem with the linker seems like an install issue with OSX [maybe picking the wrong OCaml?] For now maybe the opam package on OSX should disable native compute if the problem is so serious.

**Comment by Zimmi48** [comment]
> For now maybe the opam package on OSX should disable native compute if the problem is so serious.

If that's the solution, that's independent of the question of releasing 8.11.1, isn't it?

**Comment by ppedrot** [comment]
> If that's the solution, that's independent of the question of releasing 8.11.1, isn't it?

@Zimmi48 it probably matters for the autogenerated MacOS package? I don't know how the script works and already had a hard time with the Azure CI for the 8.11.0 release, but the flags passed to the script are likely to be part of the repo. So, I'd like to know whether it's safe to tag the release today or if we need to tweak the script options first.

**Comment by Zimmi48** [comment]
@ppedrot If the issue is only with OCaml 4.08+ then we just have to ensure that we use OCaml 4.07 or lower to generate the macOS package.

**Comment by Zimmi48** [comment]
Actually we use OCaml 4.10, so we should downgrade.

**Comment by Zimmi48** [comment]
In fact not, I was confusing with the other performance issues.

**Comment by Blaisorblade** [comment]
@Zimmi48 Will you downgrade to, say, 4.07.1 (maybe with flambda) as in https://coq.discourse.group/t/install-notes-on-coq-and-ocaml-versions-configuration/713, or can you delegate that the right way? I can also file a separate issue if that helps...

**Comment by ppedrot** [comment]
If I read correctly, native compilation is already deactivated in the MacOS package, right?

https://github.com/coq/coq/blob/c5c8ce135606f311834d4c0b9ac3e72be5ee4a36/azure-pipelines.yml#L83

**Comment by Zimmi48** [comment]
Indeed, that seems like it.

**Comment by ppedrot** [comment]
OK, so since OCaml 4.10 is fine w.r.t. performance for 8.11.1 as explained by @ejgallego in his post, let's tag.

**Comment by ejgallego** [comment]
> OK, so since OCaml 4.10 is fine w.r.t. performance for 8.11.1 as explained by @ejgallego in his post, let's tag.

Sorry, I likely wrote something confusing; but 4.10 still has significantly less performance than 4.07 in some cases; it is the best performing OCaml version >= 4.08, but if you can use 4.07.1 that's better.

Also, 4.08 doesn't allow anymore the buggy linking setup that `coq_makefile` uses for plugins so that could be a problem for some plugin users.

**Comment by ppedrot** [comment]
OK, so maybe we can generate the OSX package with a specific patch atop of the tagged commit.

**Comment by ejgallego** [comment]
Sounds good; not a big deal IMO.

**Comment by Blaisorblade** [comment]
Can native-compute be disabled by default on the command line?
If https://github.com/coq/coq/pull/12013 does that, can it be ported to 8.12 as well? That would make this issue _more_ fixed, and enable installing Coq without patching the opam scripts.

**Comment by Zimmi48** [comment]
What is the plan to close this issue exactly? It has been postponed a number of times and it is not clear to me what should be addressed in 8.12.0.

**Comment by ppedrot** [comment]
I don't think this is a Coq issue per se. Deactivating the auto-linking is merely a workaround, the real issue is that OCaml is horribly slow at linking on OSX. I don't know whether this has been reported upstream already.

**Comment by Zimmi48** [comment]
I hope that's the case, if not someone should do it. In any case, this seems very much unrelated to our 8.12.0 release process at this point, so I'm proceeding to remove the milestone.

**Comment by Blaisorblade** [comment]
Maybe consider #12564 for 8.12 instead? It's hard to set `-native-compiler` correctly without knowing what they do. Maybe the behavior of those flags should also change, but it's hard to tell.

**Comment by ppedrot** [comment]
I've just reported this as https://github.com/ocaml/ocaml/issues/9705.

**Comment by ejgallego** [comment]
> the real issue is that OCaml is horribly slow at linking on OSX

Do we know this for sure? I'd be surprised if this is the culprit; it seems to me that there is a bad linker setup arising somewhere instead.

**Comment by ejgallego** [comment]
> it seems to me that there is a bad linker setup arising somewhere instead.

Note indeed that in the extreme slowdown reports, the syscall trace is suspicious; moreover, I still believe the lazy loading should be implemented, as you don't want to pay the dynlink price [even in the case where dlopen is working fine, this is still expensive].

**Comment by Zimmi48** [comment]
Did we identify since when this slowness issue appeared? A user recently mentioned upgrading from 8.9 to 8.12 on opam and perceiving a general slowness in CoqIDE (cf. https://coq.zulipchat.com/#narrow/stream/237662-VsCoq-devs.20.26.20users/topic/Search.2FSearchAbout/near/205339604).

(I'm not sure why native compute was not disabled for macOS for version 8.12.0 on opam, cf. https://github.com/ocaml/opam-repository/pull/16876#issuecomment-665587455)

**Comment by spitters** [comment]
What's the status of this issue?

**Comment by Blaisorblade** [comment]
FTR, I'm just hearing on IRC a potential instance of a similar bug on NetBSD from @azarens-git, which is OCaml-version dependent. No action required yet.

**Comment by azarens-git** [comment]
## Observations

I've observed a similar behaviour with excessive memory consumption when I have been using ocaml compiler with spacetime profiling enabled to build coq tag V8.13.0.

This was confirmed in the pkgsrc builds on NetBSd with PKG_OPTIONS.ocaml="flambda pic spacetime", PKG_OPTIONS.ocaml="-flambda -pic -spacetime" PKG_OPTIONS="flambda pic -spacetime".  Every time the spacetime option was enabled, I got excessive memory consumption (about 100Gb residential) when coqc had been compiling theories/ (e.g. ConstructiveCauchyAbs.v, HexadecimalQ, etc.).  The compilation itself has also been slow.  Every time the spacetime option was disabled, the memory consumption has been moderate (6-8 Gb per process), and reasonably fast.

I got similar results with building coq through opam, using switches 4.11.1, 4.11.1+flambda, and 4.11.1+spacetime.  Opam builds were performed on both a Debian box and NetBSD box.

The conclusions are also supported by my observation of build process through htop.

Note: all the aforementioned builds were performed with dune.  E.g., for opam case the following instruction was used (through `opam pin edit`):

```
opam-version: "2.0"
maintainer: "coqdev@inria.fr"
authors: "The Coq development team, INRIA, CNRS, and contributors."
homepage: "https://coq.inria.fr/"
bug-reports: "https://github.com/coq/coq/issues"
dev-repo: "git+https://github.com/coq/coq.git"
license: "LGPL-2.1-only"
synopsis: "Formal proof management system"
description: """
The Coq proof assistant provides a formal language to write
mathematical definitions, executable algorithms, and theorems, together
with an environment for semi-interactive development of machine-checked
proofs. Typical applications include the certification of properties of programming
languages (e.g., the CompCert compiler certification project and the
Bedrock verified low-level programming library), the formalization of
mathematics (e.g., the full formalization of the Feit-Thompson theorem
and homotopy type theory) and teaching.
"""

depopts: [
  "coq-native"
]
depends: [
  "ocaml" {>= "4.05.0"}
  "ocamlfind" {build}
  "num"
  "conf-findutils" {build}
  "zarith" {>= "1.10"}
]
build: [ "dune" "build" "-p" "coq"]

extra-files: [
   ["coq.install" "sha512=b501737b4dbd22adc1c0377d744448056fb1dc493caf72c05f57c8463cf23f758373605ab3a50b9f$
]

url {
  src: "https://github.com/coq/coq/archive/V8.13.0.tar.gz"
}
name: "coq"
version: "8.13.0"
```

## Additional references

Additional tests were performed on the resulting `coqtop` binary executable artefact.  For more info, check out these notes https://github.com/coq/coq/issues/7829#issuecomment-777217893

The test described in that comment (involving the feeding of the script to coqtop) was performed for all builds, and the results were consistent with the previous notes; specifically, the long delays were experienced for spacetime builds, and not for the other ones.

**Comment by Blaisorblade** [comment]
With the above script, it’s unclear whether native-compiler was enabled (my first suspicion); the dune script for coq 8.13.0 at least intend to disable it. If they succeed, this might be an entirely separate issue with spacetime.
